src/tiled/tileseteditor.cpp

```diff
@@ -237,6 +237,8 @@ TilesetEditor::TilesetEditor(QObject *parent)
     connect(mTerrainDock, &TerrainDock::currentTerrainChanged, this, &TilesetEditor::currentTerrainChanged);
     connect(mTerrainDock, &TerrainDock::addTerrainTypeRequested, this, &TilesetEditor::addTerrainType);
     connect(mTerrainDock, &TerrainDock::removeTerrainTypeRequested, this, &TilesetEditor::removeTerrainType);
+    connect(mTerrainDock, &TerrainDock::moveTerrainTypeUpRequested, this, &TilesetEditor::moveTerrainTypeUp);
+    connect(mTerrainDock, &TerrainDock::moveTerrainTypeDownRequested, this, &TilesetEditor::moveTerrainTypeDown);
```

Comment:
I believe you could just implement these actions in the `TerrainDock` instead of routing them to the `TilesetEditor`. The terrain dock also has the document to which to push the undo command (`qobject_cast<TilesetDocument*>(mDocument)`).

Reply:
Good call. When I was initially writing this, I definitely felt like I was going through one too many layers of indirection.